File: re.rb
def coin(amount, array)
  return 0 if amount < 0 || array.empty?
  return 1 if amount == 0
  coin(amount - array[0], array) + coin(amount, array[1..array.size - 1])
end

p coin(1000, [1, 2, 5])

File: re.ex
defmodule Test do
  def coin_com(amount, array) do
    case {amount, array} do
      {0, ar} -> 1
      {am, ar} when am < 0 == 0 -> 0
      {am, ar} when length(ar) == 0 -> 0
      {am, [ head | tail ]} -> coin_com(am-head, [ head | tail ]) + coin_com(am, tail)
    end
  end

  Test.coin_com(1000, [1,2,5])
end
